File: Problems/Algorithms/46. Permutations/Permutations.kt
class Solution {    
    val curr: MutableList<Int>
    val res: MutableList<MutableList<Int>>
    
    init {
        curr = mutableListOf()
        res = mutableListOf()
    }
    
    fun permute(nums: IntArray): List<List<Int>> {
        val used = BooleanArray(nums.size) { false }
        aux(nums, used)
        return res
    }
    
    private fun aux(nums: IntArray, used: BooleanArray): Unit {
        if (nums.size == curr.size) {
            res.add(ArrayList(curr))
            return
        }
        
        for (i in 0..nums.size-1) {
            if (used[i]) continue
            
            used[i] = true
            curr.add(nums[i])
            aux(nums, used)
            used[i] = false
            curr.removeAt(curr.size-1)
        }
    }
}

File: Problems/Algorithms/46. Permutations/permutations.rs
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut results: Vec<Vec<i32>> = Vec::new();
        let mut curr = vec![];
        
        let mut used = vec![false; nums.len()];
        Self::aux(&nums, &mut results, &mut curr, &mut used);
        
        results
    }
    
    pub fn aux(nums: &Vec<i32>, results: &mut Vec<Vec<i32>>, curr: &mut Vec<i32>, used: &mut Vec<bool>) {
        if nums.len() == curr.len() {
            results.push(curr.clone());
            return
        }
        
        for i in 0..=nums.len()-1 {
            if used[i] {
                continue;
            }
            
            used[i] = true;
            curr.push(nums[i]);
            Self::aux(nums, results, curr, used);
            curr.pop();
            used[i] = false;
        }
    }
}

